cpp/src/fil/internal.cuh

```diff
@@ -266,6 +295,122 @@ struct forest_params_t {
 /// FIL_TPB is the number of threads per block to use with FIL kernels
 const int FIL_TPB = 256;
 
+const uint32_t max_precise_int_float = 1 << 24;  // 16'777'216
+
+__host__ __device__ __forceinline__ bool fetch_bit(const uint8_t* array, int bit)
+{
+  const int bits_per_byte = 8;
+  return array[bit / bits_per_byte] & (1 << bit % bits_per_byte);
+}
+
+struct cat_feature_counters {
+  int max_matching = -1, n_nodes = 0;
+  void reduce_with(cat_feature_counters b)
```

Comment:
It's better to have it as a standalone function or a static member function with two parameters (none of which are modified) and returning the new value.